### PR TITLE
rustc_privacy: Cleanups and perf improvements to `EmbargoVisitor`

### PR DESCRIPTION
**Pull request by petrochenkov**
Continuation of https://github.com/rust-lang/rust/pull/110907.

### CODE REVIEW

**Comment by petrochenkov** [comment]
cc @Bryanskiy

File: compiler/rustc_privacy/src/lib.rs [hunk 1]
```diff
@@ -684,75 +681,26 @@ impl<'tcx> Visitor<'tcx> for EmbargoVisitor<'tcx> {
...
+            | hir::ItemKind::ExternCrate(..)
+            | hir::ItemKind::GlobalAsm(..) => {}
+            // The interface is empty, and all nested items are processed by `visit_item`.
+            hir::ItemKind::Mod(..) => {}
             hir::ItemKind::Macro(ref macro_def, _) => {
                 self.update_reachability_from_macro(item.owner_id.def_id, macro_def);
             }
-            hir::ItemKind::ForeignMod { items, .. } => {
-                for foreign_item in items {
-                    self.update(foreign_item.id.owner_id.def_id, item_ev, Level::Reachable);
```
Comment (cjgillot) [inline compiler/rustc_privacy/src/lib.rs]: Why isn't this update needed any more?

File: compiler/rustc_privacy/src/lib.rs [hunk 2]
```diff
@@ -684,75 +681,26 @@ impl<'tcx> Visitor<'tcx> for EmbargoVisitor<'tcx> {
...
                 );
 
-                self.update_eff_vis(item.owner_id.def_id, impl_ev, None, Level::Direct);
+                if let Some(impl_ev) = impl_ev {
+                    self.update_eff_vis(item.owner_id.def_id, impl_ev, None, Level::Direct);
+                }
                 impl_ev
             }
             _ => self.get(item.owner_id.def_id),
         };
```
Comment (cjgillot) [inline compiler/rustc_privacy/src/lib.rs:682]: Can this match be merged into the following one too?

File: compiler/rustc_privacy/src/lib.rs (hunk 1 above)
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs]: It wasn't needed previously either, both nominal and effective visibilities are meaningless for `extern { ... }` items themselves, and they are not added to the table - https://github.com/rust-lang/rust/blob/master/tests/ui/privacy/effective_visibilities.rs#L10

File: compiler/rustc_privacy/src/lib.rs (hunk 2 above)
Comment (petrochenkov) [reply compiler/rustc_privacy/src/lib.rs:682]: Yes, looks like that.

**Comment by petrochenkov** [comment]
https://github.com/rust-lang/rust/pull/110907 is being reverted in https://github.com/rust-lang/rust/pull/111371 due to regressions, so marking this as blocked again.

**Comment by nnethercote** [comment]
The improvement here greatly outweigh the regressions. Nice results!

@rustbot label: +perf-regression-triaged

**Pull request merged**